File: test/imageDisplayer.lua
-- Prototype for an image displayer in lua with lua-vips
local terminal = require("preview.test.terminal")
local succes, vips = pcall(require, "vips")

local Image = {
    enabled = succes,
    instances = {},
    next_id = 1,
    cell_size = terminal.get_cell_size()
}

function Image.load(path, width, height)
    if not Image.enabled then return end

    local image = vips.Image.new_from_file(path)
    local img_width, img_height = image:size()
    local box_width, box_height = width * Image.cell_size.x, height * Image.cell_size.y

    if img_width > box_width or img_height > box_height then
        local scale = math.min(box_width / img_width, box_height / img_height)
        image = image:resize(scale)
    end

    local tmp_path = "/tmp/tty-graphics-protocol" .. Image.next_id .. ".png"
    image:write_to_file(tmp_path)

    local keys = { a = "t", t = "f", f = 100, i = Image.next_id, q = 2 }
    terminal.send_graphics_command(keys, tmp_path)

    Image.instances[path] = {
        id = Image.next_id,
        width = width,
        height = height,
    }

    Image.next_id = Image.next_id + 1

    return Image.instances[path]
end

function Image.display(path, x, y, width, height)
    if not Image.enabled then return end

    local image = Image.instances[path]
    if image == nil or image.width ~= width or image.height ~= height then
        image = Image.load(path, width, height)
    end

    local keys = { a = "p", i = (image and image.id), q = 2, C = 1 }
    terminal.move_cursor(x + 1, y)
    terminal.send_graphics_command(keys)
    terminal.restore_cursor()

    return true
end

function Image.clear(id)
    if not Image.enabled then return end

    local keys = { a = "d", d = "i", i = id }
    terminal.send_graphics_command(keys)
end

return Image

File: lib/imageDisplayer.py
import sys
import fcntl
import array
import base64
import termios
from contextlib import contextmanager

PROTOCOL_START = b"\x1b_G"
PROTOCOL_END = b"\x1b\\"
STDBOUT = getattr(sys.stdout, "buffer", sys.stdout)


@contextmanager
def temporarily_moved_cursor(x: int, y: int):
    STDBOUT.write(b"\x1b[s")
    STDBOUT.write(f"\x1b[{y+1}:{x+1}H".encode("ascii"))
    yield
    STDBOUT.write(b"\x1b[u")


def get_cell_size() -> tuple:
    buf = array.array("H", [0, 0, 0, 0])
    fcntl.ioctl(sys.stdout, termios.TIOCGWINSZ, buf)

    rows, cols, x_px_tot, y_px_tot = buf

    return x_px_tot // cols, y_px_tot // rows


def send_graphics_command(keys: dict, payload: str = "", size: int = 4096):
    data = base64.standard_b64encode(payload.encode("ascii"))
    cmd = (",".join(f"{k}={v}" for k, v in keys.items()) + ",").encode("ascii")

    for i in range(0, max(1, len(data)), size):
        cmd = cmd + (b"m=1;" if i + size < len(data) else b"m=0;")
        STDBOUT.write(PROTOCOL_START + cmd + data[i : i + size] + PROTOCOL_END)
        cmd = ""


def clear():
    keys = {"a": "d", "d": "a"}
    send_graphics_command(keys)


def load(path: str, id: int, width: int, height: int):
    try:
        from PIL import Image
    except ImportError:
        exit(1)

    image = Image.open(path)

    cell_width, cell_height = get_cell_size()
    box_size = (int(width) * cell_width, int(height) * cell_height)

    if image.width > box_size[0] or image.height > box_size[1]:
        scale = min(box_size[0] / image.width, box_size[1] / image.height)
        image = image.resize(
            (int(scale * image.width), int(scale * image.height)),
            Image.Resampling.LANCZOS,
        )

    image_name = f"/tmp/tty-graphics-protocol{id}.png"
    image.save(image_name, format="png", compress_level=0)

    keys = {"a": "t", "t": "t", "f": 100, "i": id, "q": 2}
    send_graphics_command(keys, image_name)


def display(id: int, x: int, y: int):
    keys = {"a": "p", "i": id, "q": 2}
    with temporarily_moved_cursor(x, y):
        clear()
        send_graphics_command(keys)


args = sys.argv[1:]
command = args[0]

if command == "load":
    path, id, width, height = args[1], args[2], int(args[3]), int(args[4])
    load(path, id, width, height)
elif command == "display":
    id, x, y = args[1], int(args[2]), int(args[3])
    display(id, x, y)
elif command == "clear":
    clear()
